### ISSUE TITLE
Fix more C warnings

### ISSUE THREAD

**Issue by schwehr** [HDFGroup/hdf4#32]
There are still a lot of folks using hdf4 around (I'm one of them).  I've seen a few fuzzer bugs in prior versions and am having trouble getting all the platforms I have to support working with the version at head (e.g. x86-64, ppc, and arm).  Before trying to dig into more, I figured I'd start with my current default compiler and see if I can do some low risk code cleanup if folks are up for it.  My initial run as seeing what compiler warnings are showing up:

> rm -rf build-ninja; time (mkdir -p build-ninja && cd build-ninja && cmake -DCMAKE_ANSI_CFLAGS:STRING="-Wall -Wextra -Werror -Wno-implicit-fallthrough -Wno-address -Wno-sign-compare -Wno-pedantic -Wno-stringop-truncation -Wno-type-limits -Wno-use-after-free -Wno-pointer-to-int-cast -Wno-pedantic -Wno-strict-aliasing -Wno-parentheses -Wno-tautological-compare -Wno-unused-parameter -Wno-strict-aliasing -Wno-discarded-qualifiers -Wno-implicit-function-declaration -Wno-alloc-size-larger-than -Wno-int-to-pointer-cast -Wno-array-bounds -Wno-unused-function -Wno-no-unused-const-variable= -Wno-unused-variable -Wno-memset-elt-size -Wno-maybe-uninitialized -Wno-switch -Wno-format-overflow -Wno-unused-but-set-variable" -GNinja .. && cmake --build .  && ctest -V .)

I had trouble with being able to turn off `-pedantic`, so I commented that out in https://github.com/HDFGroup/hdf4/blob/31f75f7123037768f4bc1b4ee48321316e1eaf5b/config/cmake/HDFCompilerFlags.cmake#L159

And here here in https://github.com/HDFGroup/hdf4/blob/31f75f7123037768f4bc1b4ee48321316e1eaf5b/config/cmake/HDFCompilerFlags.cmake#L159

I will try to make some manageable pull requests that just address one of those compiler warnings at a time (as I have time).

Also, I see reference to things like VAX and `$Id` that are easy cleanups.

**Comment by schwehr** [comment]
I see a number of ways to quickly improve the HDF4 codebase so that it can live as long as possible with the minimum of future upkeep work.  A lot of this stuff was done to GDAL and some to shapelib, mb-system, etc.  e.g. https://github.com/OSGeo/shapelib/issues/17 And I see that there has been some start to a few of these

- codespell - fix the many hundreds of spelling mistakes
- clang-format
- Remove unused variables
- Include what you use (iwyu)
- Order includes
- Remove comments that are no longer relevant
- Remove version history messages from file comments
- Localize variables
- Combine definition and initialization
- Add const when possible
- int boolean → bool (using [stdbool.h](https://manpages.ubuntu.com/manpages/kinetic/en/man7/stdbool.h.7posix.html))
- [stdint.h types](https://manpages.ubuntu.com/manpages/kinetic/en/man7/stdint.h.7posix.html)
- Reorder functions to not need local prototypes
- K&R prototypes to ANSI C prototypes (e.g. https://invisible-island.net/cproto/cproto.html and the now dead protoiize)
- Remove much of the code behind various debug macros. If needed, we can make some additional unit tests
  - AN_DEBUG CDEBUG CHK_DEBUG CHK_DEBUG_1 CHK_DEBUG_10 CHK_DEBUG_2 CHK_DEBUG_3 CHK_DEBUG_4 CHK_DEBUG_5 DEBUG DEBUGGING DEBUG_HDF DISKBLOCK_DEBUG EDEBUG MALDEBUG MCACHE_DEBUG MDEBUG RDEBUG SDDEBUG SYNCDEBUG VDEBUG XDRDEBUG
- Remove most of these `#ifdef` blocks:
  - added_by_mistake, DONT_KNOW, FASTER_BUT_DOESNT_WORK, FOO, LATER, MALLOC_CHECK, NOT_IMPLEMENTED, NOTNEEDED, NOTNOW, NOT_USED, NOTUSED, NOT_WORKING, NOT_YET, not_yet_implemented, OLD_WAY, TEST*, __TURBOC__
- Stop testing against `FALSE` and `TRUE`
- Remove unnecessary goto done 
-  and ?

These can be a lot smaller:

```
find . -name "*.[chf]" | xargs wc -l | grep total
 273734 total
```

**Comment by schwehr** [comment]
After looking at the build system and CI setup, it looks like hdf4 would benefit from:

- Dropping autoconf.  Supporting only one build system will simplify documentation and remove the maintenance burden 
- Define the horizon of compilers that the project will support.  e.g. Any compiler older than some threshold is explicitly not supported (maybe 5 or 7 years after release)
- Setting up a wider range of CI targets. I see that there are commented out targets.  Get those live and tune things to have better coverage.  Some examples of projects with very diverse CI runs (hdf4 doesn't probably need quite that extreme)
  - https://github.com/libgeos/geos/blob/main/.github/workflows/ci.yml
  - https://github.com/OSGeo/gdal/tree/master/.github/workflows
  - https://github.com/OSGeo/PROJ/tree/master/.github/workflows
- Use [clang static analyzer](https://clang-analyzer.llvm.org/) and Coverity
- Use [OSS Fuzz](https://google.github.io/oss-fuzz/) and [CI Fuzz](https://google.github.io/oss-fuzz/getting-started/continuous-integration/)

**Comment by schwehr** [comment]
More cleanup ideas

- Remove all uses of register.  They generally not needed with modern compilers

**Comment by derobins** [comment]
Definitely. We did that in HDF5 a while back. Should be done in HDF4, too.

Sucks that I had to revert that one PR with the silent rules in it. Need to figure out an issue we had on a creaky old machine that we have to support. Hope to get that back in later this week.

For my next trick, I'm going to start turning all the hard-coded machine configurations into feature checks. I'm hopeful that will simplify the MacOS M1/M2 situation.

Dana

**Comment by derobins** [comment]
We'll be making many of these changes over the next year or so in an effort to make HDF4 into a more maintainable piece of software. Some of them will make it into 4.2.16, but others will have to wait until later in the year, as we get 4.2.17 together.

**Comment by schwehr** [comment]
Just a note that I've seen **massive** amounts of cleanup in the last couple weeks.  There is alway always more that can be done, but the code is a lot easier to work with.  One of the biggest improvements so far is removing a lot of the dead code.
